klemens_basic_thoughts_1.md

```diff
@@ -0,0 +1,206 @@
+concept bool ProcessInitializer = requires(Init initializer, Launcher launcher) {
+        {initializer.on_setup(launcher)}   -> void;
+        {initializer.on_success(launcher)} -> void;
+        {initializer.on_error(launcher, std::error_code())} -> void;
+    };
+```
+
+A vendor implementation may provide more customization points, such as handling fork etc. Note however, that it is required to invoke the three given one from the launching process.
+
+## ProcessStream
+
+Anything that contains native_handle (HANDLE* on windows & int on linux) that can be used as a stream. Includes TCP Ports btw.
+
+# Launching a process
+
+## Core function
+
+Launching a process is done by default through the destructor of `std::process`. There are two modes to launch a process in theory: providing a `path` & arguments, OR to provide a command string. The latter will be interpreted by a shell (cmd.exe or /bin/sh). This is considered unsafe (because shell injection), and `std::system` does exactly that at the moment. Point of discussion: does that make sense to provide? Or should be move to deprecate `std::system` since we then would have a process library?
+
+In effect the two are the same in boost::process:
+
+```cpp
+system("foo && bar > file");
+system("/bin/sh", {"/c", "foo && bar > file"});
+```
+
+I will leave out the cmd-syntax for brevity sake, so here's how launching would look:
+
+```cpp
+process::process(const std::filesystem::path &path, Container args, ProcessInitializer... init);
+process::process(const std::filesystem::path &path, Container args, ProcessLauncher customLauncher, ProcessInitializer... init);
+```
+
+## Other functions
+
+In boost.process there are more functions to launcha process. After further consideration I would move to remove them all, since they are easily done by just using the process object.
+
+```cpp
+system(...) => process(...).wait()
+spawn(...) => process(...).detach()
+async_system(..., Handler, io_context) = async_wait(process(...), Handler, io_context)
+```
+
+# Classes
+
+## std::process
+
+Largely the same as `boost::process::child`.
+
+## std::process_group
+
+Similar to `boost::process::group`, BUT not used in initialization (necessarily at least). There is however a slight difference between launching a process in attached mode or attaching it later, so here's the idea:
+
+```cpp
+//launch a process attached
+process_group::emplace(const std::filesystem::path &path, Container args, ProcessInitializer... init); //returns either void or pid_t -> point of discussion
+process_group::emplace(process::process(const std::filesystem::path &path, Container args, ProcessLauncher customLauncher, ProcessInitializer... init);
+
+//attach an existing process
+process_group::attach(process && p); //meaning p won't have a handle on the subprocess anymore
+
+//detach a process who's pid I know
+process_group::detach(pid_t pid) -> child; 
+```
+
+## Pipe & async_pipe & pstream
```

Comment:
so during LIAW we were thinking we should just push all async behavior into a 'maybe later' or not specified bucket -- since there's so much going on in the committee.  Also simplifies the first paper.  Would you be ok with that?

Reply:
I am not sure about this one. The async_pipe are almost necessary for actual use (so a process-lib without them is not a good idea). However, getting the async-pipe in would be easy after we get the Networking-TS, since it is using the same mechanics as sockets.

The async-waiting is a bigger deal, because that is different. We could seperate it into a different paper, but I don't think a Std::process without is that awesome. Just like a Networking lib that's only sync wouldn't be a good choice, while still usable to some.

Reply:
Agree -- I think some of the workshop participants were afraid that would get the proposal embroiled in the executors and networking TS -- which is the controversy of the day at the committee.  I'm of the mind we propose it and be clear that our intent is to integrate with the solution.